crypto/mlkem/mlkem-all.h

```diff
@@ -0,0 +1,73 @@
+/* Variant-neutral helper functions */
+
+#include <crypto/mlkem.h>
+
+typedef mlkem_scalar scalar;
+typedef ossl_mlkem_ctx mctx;
+typedef mlkem_cbd_func cbd_t;
+typedef const mlkem_vinfo *vinfo_t;
+
+/*
+ * These are global symbols in the static (.a) library, so need "ossl_mlkem_",
+ * or similar, prefixes.
+ */
+
+/*
+ * Variant-specific CBD vector generation helpers, these generate a single CBD
+ * scalar.
+ */
+__owur cbd_t ossl_mlkem_cbd_2;
+__owur cbd_t ossl_mlkem_cbd_3;
+
+/*
+ * These functions implement the internals of the variant-specific API.
+ *
+ * All vectors and matrices are passed as a pointer to their first scalar
+ * element, the "vinfo" |rank| value determines the dimensions.
+ *
+ * Where temporary vectors are needed the caller passes an appropriately sized
+ * mutable object.
+ */
+void ossl_mlkem_encode_public_key(
+        uint8_t *out, const scalar *t, const uint8_t rho[MLKEM_RANDOM_BYTES],
+        vinfo_t vinfo);
```

Comment:
Nit: If this is a "public" API prefixed "ossl_" shouldn't then also all parameter types be "public" (and also suitably be prefixed "ossl_")? In this case, "ossl_mlkem_vinfo"?

Reply:
If the function is public and ossl_ prefixed any typedefs it requires should also be prefixed in the same manner.
We aren't consistent on that - but we should not dig the hole deeper.

And in particular this one should be for sure.

Reply:
No, this is not actually a "public" API.  Rather, it is a global symbol when the library is a ".a", but we'll not provide any headers or documentation for calling these.

Reply:
That's understood @vdukhovni -- I just wondered about consistency again. But I don't have a strong opinion on this one; shorter actually is more beautiful :) So close this comment as you see fit.

Reply:
While we're discussing naming, one last time (review  feedback aside) on my plat is to add a public header file with just the externally observable constants that are seen during encapsulation and key management. So something like:

```
#define OSSL_MLKEM_SHARED_SECRET_BYTES      32

#define OSSL_MLKEM512_SECURITY_BITS         128
#define OSSL_MLKEM512_CIPHERTEXT_BYTES      768
#define OSSL_MLKEM512_PUBLIC_KEY_BYTES      800
#define OSSL_MLKEM512_PRIVATE_KEY_BYTES     1632

#define OSSL_MLKEM768_SECURITY_BITS         192
#define OSSL_MLKEM768_CIPHERTEXT_BYTES      1088
#define OSSL_MLKEM768_PUBLIC_KEY_BYTES      1184
#define OSSL_MLKEM768_PRIVATE_KEY_BYTES     2400

#define OSSL_MLKEM1024_SECURITY_BITS        256
#define OSSL_MLKEM1024_CIPHERTEXT_BYTES     1568
#define OSSL_MLKEM1024_PRIVATE_KEY_BYTES    3168
#define OSSL_MLKEM1024_PUBLIC_KEY_BYTES     1568
```

Modulo various additional `_` characters (or not). And of course the "security bits" are not really a constant, things change, these are just the currently claimed and somewhat contested numbers, so likely would not actually go in, but there mostly to suggest that perhaps I'd forgotten something else to list, though it seems like like knowing these is enough for doing key generation encap/decap.

Reply:
The NIST Security Category value should also be there - and be able to be retrieved via params.

Basically that is their defined 1-5 mapping value and meant to replace the concept of using something like security bits to make life easier for those who get confused about bits in different contexts.

i.e. these things:

• ML-KEM-512 (security category 1)
• ML-KEM-768 (security category 3)
• ML-KEM-1024 (security category 5)

Reply:
The "bits" security bits are available via params, I am not aware of which parameters if any correspond to the `1–5` security levels.  It was actually quite difficult to find these defined by NIST, their reference to SP800-57 did not lead do any such definition, however the PDF of the PQ call for proposals was one place where I ultimately found these defined.

Anyway, if you'd like to see specific things happen before this is initially merged (subject to any further refinement before release, and as part of FIPS module integration, ....) can you list exceedingly precise changes to be considered now, before I go spending time on getting everything renamed...  There would probably need to be some sign of support or at least acquiescence for/to the proposal from others, I don't want to do this twice.

Reply:
Don't worry, the additional values are a mere digression. The main thing now is code review, and **if** the (internal only!, no public API) symbol names of the implementation MUST change, this should be decided promptly and have some measure of rough consensus.

Reply:
Agreed, the security category is separate from what this PR is doing.

I'd suggest that it should be it's own independent feature branch that adds the security category for **all** algorithms.  It might not be large enough for a feature branch but I think it is.

Reply:
A note re the original question in this thread: it has been argued, a number of years ago, that function parameter names should be made "unique" in some way, to void clashes with possibly common names (what if someone, somewhere, has a macro called `vinfo`, for example).

So yes, with that in mind, it's quite possible that `vinfo` → `ossl_vinfo` could be a good move.

That being said, in spite of the discussion a number of years ago, we didn't make this a habit.  Perhaps now's a good time to start?

Reply:
> So yes, with that in mind, it's quite possible that `vinfo` → `ossl_vinfo` could be a good move.
> 
> That being said, in spite of the discussion a number of years ago, we didn't make this a habit. Perhaps now's a good time to start?

The parameter name problem is really about hygiene with *named* parameters in header files, which may be used in applications that import random other headers, and where conflict potentially becomes an issue. The standard approach used to avoid that problem is to **NOT** include the parameter names in function *prototypes*. I strongly support that approach, but recall failing to convince the team quite some years ago.

Meanwhile, parameters used **inside** our own C-code modules are not subject to the same concerns, and we don't have to worry about function parameters named `t`, `m`, `y`, `s`, ..., or types named `vinfo_t` which is actually a file-scoped `typedef` for what is really `const mlkem_vinfo *`.

Reply:
> > ... The standard approach used to avoid that problem is to **NOT** include the parameter names in function _prototypes_.
> 
> Sure, that works too... but you lose the slight documentation that the name provides (especially with generic types). Gotta remember to add those back in form of comments.

Yes, the actual parameter names are then **documented**, and one does not get to be negligent by pretending that the `.h` files are *documentation*. Sure, comments in the `.h` files can also be helpful when someone is already knowledgeable, who's familiar with what the functions do, ...,  and just does not recall the *synopsis*, so if the parameter names are removed from the prototypes, by all means provide them in comments above each prototype.

Reply:
... or in the prototype:

``` C
void ossl_mlkem_encode_public_key(
        uint8_t * /* out */, const scalar * /* t */, const uint8_t /* rho */[MLKEM_RANDOM_BYTES],
        vinfo_t /* vinfo */);
```

Two reasons:

- when you grep for the function name, you get the names served in place, no need to go look *somewhere else* for the documentation when the name could be enough.
- documentation elsewhere may be out of date, while the names in place are more likely to be up to date.  This is an observation of a fairly common (not just us) documentation problem.